src/datachain/lib/memory_utils.py

```diff
@@ -0,0 +1,57 @@
+"""Memory estimation utilities for DataChain."""
+
+import sys
+from typing import Any, Union
+
+# Default batch processing values
+DEFAULT_CHUNK_ROWS = 2000
+DEFAULT_CHUNK_MB = 1000
+
+# Memory monitoring threshold (percentage)
+MEMORY_USAGE_THRESHOLD = 80
+
+# System memory check frequency (every N rows)
+MEMORY_CHECK_FREQUENCY = 100
+
+# Shared constant for object overhead estimation
+OBJECT_OVERHEAD_BYTES = 100
+
+
+def estimate_memory_recursive(item: Any) -> int:
+    if item is None:
+        return 0
+
+    if isinstance(item, (str, bytes, int, float, bool)):
+        return sys.getsizeof(item)
+    if isinstance(item, (list, tuple)):
+        total_size = sys.getsizeof(item)
+        for subitem in item:
+            total_size += sys.getsizeof(subitem)
+        return total_size
+    # For complex objects, use a conservative estimate
+    return sys.getsizeof(item) + OBJECT_OVERHEAD_BYTES
```

Comment:
This likely won't work since we mostly deal with complex objects. I can guess that logic will get here in 99% of cases.